main/JSONParameterUtilsMod.F90

```diff
@@ -0,0 +1,1304 @@
+    
+    ! ---------------------------------------------------------------------------------
+    ! Step 2: Advance through the file until the next closing bracket. Save
+    !         EVERYTHING in a large string.
+    ! ---------------------------------------------------------------------------------
+
+    found_close = .false.
+    vardata_str = ''
+    i=0
+    do_close: do while(.not.found_close)
+
+       read(unit=file_unit,fmt='(A1)',ADVANCE='NO', iostat=io_status, iomsg=io_msg) filechar
+       if (io_status == IOSTAT_END) then
+          write(log_unit,*) 'encountered EOF'
+          call shr_sys_abort()
+       elseif (io_status < 0) then
+          cycle do_close
+       elseif (io_status>0) then
+          write(log_unit,*) 'fatal i/o error! code:', io_status
+          write(log_unit,*) 'msg: ',io_msg
+          call shr_sys_abort()
+       end if
+
+       i=i+1
+       
+       if(i<=vardata_max_len)then
+          vardata_str(i:i) = filechar
+       else
+          write(log_unit,*) 'Ran out of room reading in variable data'
+          write(log_unit,*) 'Increase the size of integer constant: JSONParameterUtilsMod:vardata_max_len'
+          call shr_sys_abort()
+       end if
+       if(index(vardata_str,'{')>0)then
+          write(log_unit,*)'An open bracket was found nested inside the current parameter:',trim(symb_str)
+          write(log_unit,*)'This is an invalid file format for the parameter file'
+          call shr_sys_abort()
+       end if
+       if(index(vardata_str,'}')>0)then
+          found_close = .true.
+       end if
+    end do do_close
+    vardata_len = i
+    
+    if(phase==count_phase)then
+       var_num = var_num + 1
+       ! We've advance the file pointer to the closing bracket
+       ! if we are just counting, we are done
+       return
+    end if
+
+    ! ---------------------------------------------------------------------------------
+    ! Step 3: Read through that string. Fill data structures.
+    ! ---------------------------------------------------------------------------------
+
+    param => pstruct%parameters(var_num)
+    param%name = trim(adjustl(symb_str))
+    param%access_count = 0
+
+    if(debug) write(log_unit,*) 'Parameter: ',trim(param%name)
+    
+    call GetMetaString(vardata_str,'"dims"',beg_id,end_id)
+    call GetStringVec(vardata_str(beg_id:end_id),string_scr,n_vec_out)
+    allocate(param%dim_names(n_vec_out))
+    dimsizes(:)=-1
+    param%ndims = n_vec_out
+    do i = 1,n_vec_out
+       param%dim_names(i) = trim(CleanSymbol(string_scr(i)))
+       dimsizes(i) = pstruct%GetDimSizeFromName(trim(param%dim_names(i)))
+    end do
+    call ClearStringScratch(string_scr,n_vec_out)
+
+    
+    call GetMetaString(vardata_str,'"dtype"',beg_id,end_id)
+    call GetStringVec(vardata_str(beg_id:end_id),string_scr,n_vec_out)
+    data_str = trim(string_scr(1))
+    param%dtype = -999
+    if(trim(CleanSymbol(data_str))=='float') then
```

Comment:
The list of if's for data type, would be better to be inside a a helper function to hide the complexity having to do with type.